Move off of UIKit SPI: UIWebFormAccessory and UIWebFormAccessoryDelegate

whsieh opened the pull request:
#### edd2d7016e2d3e144e9cdf36b13da64e92b1c994
<pre>
Move off of UIKit SPI: UIWebFormAccessory and UIWebFormAccessoryDelegate
<a href="https://bugs.webkit.org/show_bug.cgi?id=262121">https://bugs.webkit.org/show_bug.cgi?id=262121</a>

Reviewed by Tim Horton.

Stop using this SPI class and delegate protocol; instead, introduce `WKFormAccessoryView` and
`WKFormAccessoryViewDelegate`, which are (nearly) drop-in replacements for `UIWebFormAccessory`.
The only functionality on `UIWebFormAccessory` which we drop in our new WebKit class is support for
showing a &quot;Clear&quot; button (`-setClearVisible:`), which was only necessary when using the &quot;wheels of
time&quot; UI in the keyboard on iPhone; this hasn&apos;t been necessary since the form control refresh,
which moved date pickers into a context menu presentation (and more recently, into popover UI).

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove now-unused SPI declarations.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKFormInputSession accessoryViewCustomButtonTitle]):
(-[WKContentView accessoryDone]):
(-[WKContentView accessoryViewDone:]):
(-[WKContentView accessoryTab:]):
(-[WKContentView accessoryView:tabInDirection:]):

Instead of using the existing `-accessoryTab:` delegate that takes a `BOOL`, we can instead use an
`enum class` to represent the direction (next or previous) for slightly better clarity.

(-[WKContentView accessoryViewAutoFill:]):
(-[WKContentView formAccessoryView]):
(-[WKContentView _updateAccessory]):
(-[WKContentView accessoryClear]): Deleted.
(-[WKContentView accessoryAutoFill]): Deleted.

Make `_formAccessoryView` a `WKFormAccessoryView` instead of a `UIWebFormAccessory`, and conform to
`WKFormAccessoryViewDelegate` instead of `UIWebFormAccessoryDelegate`.

* Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.h: Added.
* Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.mm: Added.
(WebKit::doneButtonFont):
(WebKit::singleLineAutoFillButtonFont):
(WebKit::multipleLineAutoFillButtonFont):
(WebKit::upArrow):
(WebKit::downArrow):
(WebKit::createToolbarWithItems):
(-[WKFormAccessoryView _initForUniversalControlBar:]):
(-[WKFormAccessoryView initWithInputAssistantItem:delegate:]):
(-[WKFormAccessoryView traitCollectionDidChange:]):
(-[WKFormAccessoryView _updateFrame]):
(-[WKFormAccessoryView layoutSubviews]):
(-[WKFormAccessoryView _done]):
(-[WKFormAccessoryView _previousTapped]):
(-[WKFormAccessoryView _nextTapped]):
(-[WKFormAccessoryView _autoFill]):
(-[WKFormAccessoryView _autoFillButton]):
(-[WKFormAccessoryView _refreshAutofillPresentation]):
(-[WKFormAccessoryView _applyTreatmentToAutoFillLabel]):
(-[WKFormAccessoryView hideAutoFillButton]):
(-[WKFormAccessoryView showAutoFillButtonWithTitle:]):
(-[WKFormAccessoryView autoFillButtonItem]):
(-[WKFormAccessoryView setNextPreviousItemsVisible:]):
(-[WKFormAccessoryView setNextEnabled:]):
(-[WKFormAccessoryView isNextEnabled]):
(-[WKFormAccessoryView setPreviousEnabled:]):
(-[WKFormAccessoryView isPreviousEnabled]):
(-[WKFormAccessoryView contentRatio]):
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::promptForGetDisplayMedia):

Unrelated unified source build fix.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/ios/UIKitSPI.h:

Canonical link: <a href="https://commits.webkit.org/268485@main">https://commits.webkit.org/268485@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bdedfe73f796e7db117fe6d2720c21a8b60471e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19833 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20875 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21725 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/18524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20068 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23515 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20404 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/20068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20053 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20035 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17246 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22579 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17207 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18041 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/24319 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18283 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18215 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22302 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17978 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4743 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22326 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

megangardner commented on Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.mm:
```diff
@@ -0,0 +1,432 @@
+
+ALLOW_DEPRECATED_IMPLEMENTATIONS_BEGIN
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+// FIXME: Adopt -registerForTraitChanges: once iOS 17 is the minimum supported iOS version.
```
Ditto

megangardner commented on Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.mm:
```diff
@@ -0,0 +1,432 @@
+    [self _updateFrame];
+
+    ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    // FIXME: Adopt -registerForTraitChanges: once iOS 17 is the minimum supported iOS version.
```
Can you file a radar for this?

whsieh replied:
Filed a bugzilla: <https://bugs.webkit.org/show_bug.cgi?id=262139> — I'll reference it here.

whsieh committed:
Move off of UIKit SPI: UIWebFormAccessory and UIWebFormAccessoryDelegate
https://bugs.webkit.org/show_bug.cgi?id=262121

Reviewed by Tim Horton.

Stop using this SPI class and delegate protocol; instead, introduce `WKFormAccessoryView` and
`WKFormAccessoryViewDelegate`, which are (nearly) drop-in replacements for `UIWebFormAccessory`.
The only functionality on `UIWebFormAccessory` which we drop in our new WebKit class is support for
showing a "Clear" button (`-setClearVisible:`), which was only necessary when using the "wheels of
time" UI in the keyboard on iPhone; this hasn't been necessary since the form control refresh,
which moved date pickers into a context menu presentation (and more recently, into popover UI).

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

Remove now-unused SPI declarations.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKFormInputSession accessoryViewCustomButtonTitle]):
(-[WKContentView accessoryDone]):
(-[WKContentView accessoryViewDone:]):
(-[WKContentView accessoryTab:]):
(-[WKContentView accessoryView:tabInDirection:]):

Instead of using the existing `-accessoryTab:` delegate that takes a `BOOL`, we can instead use an
`enum class` to represent the direction (next or previous) for slightly better clarity.

(-[WKContentView accessoryViewAutoFill:]):
(-[WKContentView formAccessoryView]):
(-[WKContentView _updateAccessory]):
(-[WKContentView accessoryClear]): Deleted.
(-[WKContentView accessoryAutoFill]): Deleted.

Make `_formAccessoryView` a `WKFormAccessoryView` instead of a `UIWebFormAccessory`, and conform to
`WKFormAccessoryViewDelegate` instead of `UIWebFormAccessoryDelegate`.

* Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.h: Added.
* Source/WebKit/UIProcess/ios/forms/WKFormAccessoryView.mm: Added.
(WebKit::doneButtonFont):
(WebKit::singleLineAutoFillButtonFont):
(WebKit::multipleLineAutoFillButtonFont):
(WebKit::upArrow):
(WebKit::downArrow):
(WebKit::createToolbarWithItems):
(-[WKFormAccessoryView _initForUniversalControlBar:]):
(-[WKFormAccessoryView initWithInputAssistantItem:delegate:]):
(-[WKFormAccessoryView traitCollectionDidChange:]):
(-[WKFormAccessoryView _updateFrame]):
(-[WKFormAccessoryView layoutSubviews]):
(-[WKFormAccessoryView _done]):
(-[WKFormAccessoryView _previousTapped]):
(-[WKFormAccessoryView _nextTapped]):
(-[WKFormAccessoryView _autoFill]):
(-[WKFormAccessoryView _autoFillButton]):
(-[WKFormAccessoryView _refreshAutofillPresentation]):
(-[WKFormAccessoryView _applyTreatmentToAutoFillLabel]):
(-[WKFormAccessoryView hideAutoFillButton]):
(-[WKFormAccessoryView showAutoFillButtonWithTitle:]):
(-[WKFormAccessoryView autoFillButtonItem]):
(-[WKFormAccessoryView setNextPreviousItemsVisible:]):
(-[WKFormAccessoryView setNextEnabled:]):
(-[WKFormAccessoryView isNextEnabled]):
(-[WKFormAccessoryView setPreviousEnabled:]):
(-[WKFormAccessoryView isPreviousEnabled]):
(-[WKFormAccessoryView contentRatio]):
* Source/WebKit/UIProcess/mac/DisplayCaptureSessionManager.mm:
(WebKit::DisplayCaptureSessionManager::promptForGetDisplayMedia):

Unrelated unified source build fix.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/ios/KeyboardInputTestsIOS.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/ios/UIKitSPI.h:

Canonical link: https://commits.webkit.org/268485@main
